packages/alfa-css/src/value/contain.ts

```diff
@@ -0,0 +1,237 @@
+  }
+
+  public hash(hash: Hash): void {
+    hash
+      .writeBoolean(this._size)
+      .writeBoolean(this._inlineSize)
+      .writeBoolean(this._layout)
+      .writeBoolean(this._style)
+      .writeBoolean(this._paint);
+  }
+
+  public toJSON(): ContainFlags.JSON {
+    return {
+      ...super.toJSON(),
+      size: this._size,
+      inlineSize: this._inlineSize,
+      layout: this._layout,
+      style: this._style,
+      paint: this._paint,
+    };
+  }
+
+  public toString(): string {
+    let keywords: Array<string> = [];
+    if (this._size) {
+      keywords.push("size");
+    }
+    if (this._inlineSize) {
+      keywords.push("inline-size");
+    }
+    if (this._layout) {
+      keywords.push("layout");
+    }
+    if (this._style) {
+      keywords.push("style");
+    }
+    if (this._paint) {
+      keywords.push("paint");
+    }
+
+    return keywords.join(" ");
+  }
+}
+
+/**
+ * @public
+ */
+export namespace ContainFlags {
+  export interface JSON extends Value.JSON<"contain-flags"> {
+    size: boolean;
+    inlineSize: boolean;
+    layout: boolean;
+    style: boolean;
+    paint: boolean;
+  }
+
+  export function isContainFlags(value: unknown): value is ContainFlags {
+    return value instanceof ContainFlags;
+  }
+
+  export const parse: CSSParser<ContainFlags> = (input) => {
+    let size: Keyword<"size"> | Keyword<"inline-size"> | undefined;
+    let layout: Keyword<"layout"> | undefined;
+    let style: Keyword<"style"> | undefined;
+    let paint: Keyword<"paint"> | undefined;
+
+    while (true) {
+      for (const [remainder] of Token.parseWhitespace(input)) {
+        input = remainder;
+      }
+
+      if (size === undefined) {
+        const result = Keyword.parse("size", "inline-size")(input);
+
+        if (result.isOk()) {
+          [input, size] = result.get();
+          continue;
+        }
+      }
+
+      if (layout === undefined) {
+        const result = Keyword.parse("layout")(input);
+
+        if (result.isOk()) {
+          [input, layout] = result.get();
+          continue;
+        }
+      }
+
+      if (style === undefined) {
+        const result = Keyword.parse("style")(input);
+
+        if (result.isOk()) {
+          [input, style] = result.get();
+          continue;
+        }
+      }
+
+      if (paint === undefined) {
+        const result = Keyword.parse("paint")(input);
+
+        if (result.isOk()) {
+          [input, paint] = result.get();
+          continue;
+        }
+      }
+
+      break;
+    }
+
+    if (
+      size === undefined &&
+      layout === undefined &&
+      style === undefined &&
+      paint === undefined
+    ) {
+      return Err.of("Expected at least one keyword");
+    }
+
+    return Result.of([
+      input,
+      ContainFlags.of(
+        size !== undefined && size.value === "size",
+        size !== undefined && size.value === "inline-size",
+        layout !== undefined,
+        style !== undefined,
+        paint !== undefined,
```

Comment:
**Note:** we could store these directly as boolean and use `layout ?? false` here. I'm not sure that really improve things since it means we wouldn't be able to directly do `[input, layout] = result.get()`. Or maybe by wrapping the parser in `const result = map(Keyword.parse("layout"), () => true)(input)`. But that's not really better…

Reply:
I'll leave this as is for now 🙂